enterprise/utils.py

```diff
@@ -377,6 +382,8 @@ def build_notification_message(template_context, template_configuration=None):
         template_context (dict): A set of data to render
         template_configuration: A database-backed object with templates
             stored that can be used to render a notification.
+        default_text_template: template to use for text emails, if template_configuration is not found.
+        default_html_template: template to use for html emails, if template_configuration is not found.
```

Comment:
The more I go through this PR the more I lose my mind.

Okay, so:
1. We're using 2 storage methods for templates - file and database
2. We previously (before this PR) had the "default" as a file based template because we associated the templates ONLY with an enterprise customer.
3. You're adding _another_ file based template, which is not what I want and I think we can get rid of ALL file-based templates and just consistently use the database.
4. We can leave the OneToOne logic be and just make it nullable.

So for build_notification_message, I'd actually get rid of code, not add any, it should look like this:
```
def build_notification_message(
        template_context,
        template_configuration=None):
    return template_configuration.render_all_templates(template_context)
```
And where you figure out your template configuration, the logic should just be:
```
if bulk_enrollment:
    enterprise_template_config = enterprise_customer.enterprise_enrollment_templates.filter(
            template_type=BULK_ENROLLMENT_TEMPLATE_TYPE
    ).first()
else:
    try:
        enterprise_template_config = enterprise_customer.enterprise_enrollment_template

    except (ObjectDoesNotExist, AttributeError):
        enterprise_template_config = enterprise_customer.enterprise_enrollment_templates.filter(
            template_type=DEFAULT_TEMPLATE_TYPE
        ).first()
```
Do you see what I'm getting at with this?
The database already stores html and plaintext template and subject and all that jazz, we should just add bulk enrollment and the default to that and then figure out which template to call from the database rather than make a database call and fallback to a filesystem.

Reply:
hmm I would love that too actually, except there is n issue we will hit. The issue will be that there will have to be a template model instance for every enterprise customer, and then have a way to fall back if that does not exist. If we remove the file based template, we will then need to furnish two things: an alternative fallback method (which would be another database record somewhere?) .

not sure if I clearly explained but mainly the issue is that there is only one record in prod in this table, so almost all enterprises are using the 'fallback' from files as of now.  So it's almost as if we never meant to use the db models haha... 

Now, if there was a fallback template also in the database, then I think we can make it work all database based and eliminate the file based system. Would that be a route we are interested in? Even if we do, I think we may want to do that separately not part of this PR.. thoughts?

Reply:
in other words, if we were to redo this from scratch, I think we will need: (1) a default fallback template for both types in db, and (2) a per enterprise template like today. No?

Reply:
Right so (and this is mentioned in another comment) our current logic even before bulk enrollment, is default OR customer. If bulk enrollment is just another template (which is sort of how it was being treated, your code actually breaks existing customer templates, so we need to fix that either way), then that's where my logic comes into play. It uses bulk enrollment templates first and foremost when that is specified, if not, it tries to fetch a customer template and then will fall back to the default.

In the case of us having a bulk enrollment template customized for a customer then yes this logic needs to change a little, but it still roughly follows the above.

Reply:
Ah so I see where we have a disconnect, there is a dimension I did not clarify before, which I think may clarify this better?

The thing is that we will still need to support a template per ' enrollment type'. For example:

- if bulk enrollment: use template default OR customer
- else regular enrollment: use template default OR customer

So each enrollment type will have a default template, and also a possible customer override (which I am honestly thinking is not going to be used much).

Given that, does the proposed setup then make sense?

Reply:
also my logic see below, basically does what you mentioned, but later it still uses file based fallback if this results in None.

```
try:
        if admin_enrollment:
            template_type = ADMIN_ENROLL_EMAIL_TEMPLATE_TYPE
        else:
            template_type = DEFAULT_ENROLL_EMAIL_TEMPLATE_TYPE
        # we only support one template per type
        enterprise_template_config = enterprise_customer.enterprise_enrollment_templates.filter(
            template_type=template_type
        ).first()
    except (ObjectDoesNotExist, AttributeError):
        enterprise_template_config = None
```


I think really the only moving part here is whether we use file based templates at all. Replacing that with a db based template is possible but it requires another refactor plus a new model, we cannot fit it into the current model which is designed for per customer overrides only.

Reply:
@binodpant am I correct that the full "proposed setup" after this conversation is: 
 1.  putting both defaults in the DB
      1a. Write a migration that put the defaults in the new DB immediately on deploy of this change. 
 2. Remove file-based templates completely
 3. Keeping the existing customer overrides for a learner-initiated enrollment (which we already have but we don't think is used much), and apply those under condition "learner-iniated enrollment AND customer has _default type_ template" 
 4. (this is the one I am muddiest on from the above conversation: do we need this or do all bulk-enrolled learners just get 1 email? ) Add capability to store new customer templates for the bulk-enrollment initiated enrollments. Apply this template under condition "bulk/admin-iniated enrollment AND customer has _bulk/admin type_ template"

Reply:
so resolving this, since we are now trying completely database based templates. no more files.

tests upcoming but ...